### PR TITLE
Fixes #2032

### LINKED ISSUE #2032: CCPM extension to have weights instead of reversals
Currently in the heli menu, OpenTX only allows you to reverse input channels that are inputs into the CCPM mix.  This is a request to change those from check box reversals into full weight numbers.

When programming a CCPM heli, a user can have multiple pitch curves and different DR's for aileron and elevator.  At the moment for CCPM program, these need to be re-adjusted to give input weights to the CCPM algorithm (which can be done by using global variables, but this is very clunky).

Can these reversals just be input weights percentages from -100.0 to +100.0?  All the weights do is multiply each input by the weight percentage before processing by the CCPM algorithm.  (Ail by Ail%, Ele by Ele%, Collective by Collective%).

Existing models could be translated to have weights of -100% or +100% (based on current check boxes).

For more information see open discussion on HeliFreak: http://www.helifreak.com/showthread.php?t=676944

### PR DESCRIPTION
**Pull request by bsongis**
Don't push right now, it's not ready (at least Companion part is missing). But I will need a code review!

### CODE REVIEW

File: radio/src/mixer.cpp [hunk 1]
```diff
@@ -593,15 +590,22 @@ void evalFlightModeMixes(uint8_t mode, uint8_t tick10ms)
...
+#endif
   if (g_model.swashR.value) {
-    uint32_t v = ((int32_t)HELI_ANAS_ARRAY(ELE_STICK)*HELI_ANAS_ARRAY(ELE_STICK) + (int32_t)HELI_ANAS_ARRAY(AIL_STICK)*HELI_ANAS_ARRAY(AIL_STICK));
+    uint32_t v = ((int32_t)heliEleValue*heliEleValue + (int32_t)heliAilValue*heliAilValue);
     uint32_t q = calc100toRESX(g_model.swashR.value);
     q *= q;
     if (v>q) {
       uint16_t d = isqrt32(v);
       int16_t tmp = calc100toRESX(g_model.swashR.value);
-      HELI_ANAS_ARRAY(ELE_STICK) = (int32_t) HELI_ANAS_ARRAY(ELE_STICK)*tmp/d;
```
Comment (bsongis) [inline radio/src/mixer.cpp]: This is strange to me, we wrote anas (and heliAnas) as many times as active flight modes? I think it was a bug.

File: radio/src/gui/Taranis/menu_model_heli.cpp [hunk 2]
```diff
@@ -37,52 +37,77 @@
...
 
-  for (uint8_t i=0; i<6; i++) {
-    coord_t y = MENU_TITLE_HEIGHT + 1 + i*FH;
-    uint8_t attr = (sub == i ? ((s_editMode>0) ? BLINK|INVERS : INVERS) : 0);
-    switch(i) {
-      case ITEM_HELI_SWASHTYPE:
-        g_model.swashR.type = selectMenuItem(HELI_PARAM_OFS, y, STR_SWASHTYPE, STR_VSWASHTYPE, g_model.swashR.type, 0, SWASH_TYPE_MAX, attr, event);
-        break;
+  for (unsigned int i=0; i<NUM_BODY_LINES; i++) {
+    coord_t y = MENU_HEADER_HEIGHT + 1 + i*FH;
```
Comment (projectkk2glider) [inline radio/src/gui/Taranis/menu_model_heli.cpp]: MENU_HEADER_HEIGHT is undefined

File: radio/src/gui/Taranis/menu_model_heli.cpp [hunk 3]
```diff
@@ -37,52 +37,77 @@
...
-        lcd_outdezAtt(HELI_PARAM_OFS, y, g_model.swashR.value,  LEFT|attr);
+        lcd_outdezAtt(MODEL_HELI_2ND_COLUMN, y, g_model.swashR.value, LEFT|attr);
         if (attr) CHECK_INCDEC_MODELVAR_ZERO(event, g_model.swashR.value, 100);
         break;
 
-      case ITEM_HELI_ELEDIRECTION:
-        g_model.swashR.invertELE = selectMenuItem(HELI_PARAM_OFS, y, STR_ELEDIRECTION, STR_MMMINV, g_model.swashR.invertELE, 0, 1, attr, event);
+      case ITEM_HELI_ELE:
+        lcd_putsLeft(y, "Elevator" /*STR_ELEVATOR_SOURCE*/);
+        putsMixerSource(MODEL_HELI_2ND_COLUMN, y, g_model.swashR.elevatorSource, attr);
```
Comment (projectkk2glider) [inline radio/src/gui/Taranis/menu_model_heli.cpp]: Input names have wrong ofset:
snapshot_08

File: radio/src/gui/Taranis/menu_model_heli.cpp (hunk 2 above)
Comment (bsongis) [inline radio/src/gui/Taranis/menu_model_heli.cpp]: Oh I missed some files

File: radio/src/gui/Taranis/menu_model_heli.cpp (hunk 3 above)
Comment (bsongis) [inline radio/src/gui/Taranis/menu_model_heli.cpp]: Strange I tested it!!!

Comment (bsongis) [inline radio/src/gui/Taranis/menu_model_heli.cpp]: Isn't it your EEPROM which is shifted? This commit introduces a shift in 217!

Comment (projectkk2glider) [inline radio/src/gui/Taranis/menu_model_heli.cpp]: On 14. 02. 2015 18:34, Bertrand Songis wrote:
>
> In radio/src/gui/Taranis/menu_model_heli.cpp 
> <https://github.com/opentx/opentx/pull/2057#discussion_r24715807>:
>
> >          if (attr) CHECK_INCDEC_MODELVAR_ZERO(event, g_model.swashR.value, 100);
> >          break;
> >
> > -      case ITEM_HELI_ELEDIRECTION:
> > -        g_model.swashR.invertELE = selectMenuItem(HELI_PARAM_OFS, y, STR_ELEDIRECTION, STR_MMMINV, g_model.swashR.invertELE, 0, 1, attr, event);
> > +      case ITEM_HELI_ELE:
> > +        lcd_putsLeft(y, "Elevator" /*STR_ELEVATOR_SOURCE*/);
> > +        putsMixerSource(MODEL_HELI_2ND_COLUMN, y, g_model.swashR.elevatorSource, attr);
>
> Isn't it your EEPROM which is shifted? This commit introduces a shift 
> in 217!
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/opentx/opentx/pull/2057/files#r24715807>.
>
Yes, deleted eeprom.bin, now everything OK!

File: radio/src/mixer.cpp [hunk 4]
```diff
@@ -152,17 +138,6 @@ void applyExpos(int16_t *anas, uint8_t mode APPLY_EXPOS_EXTRA_PARAMS)
...
-        if (ed->srcRaw == MIXSRC_Ele) {
-          heliAnas[ELE_STICK] = v;
-          heliTrims[ELE_STICK] = virtualInputsTrims[cur_chn];
-        }
-        else if (ed->srcRaw == MIXSRC_Ail) {
-          heliAnas[AIL_STICK] = v;
-          heliTrims[AIL_STICK] = virtualInputsTrims[cur_chn];
-        }
-#endif
 #endif
```
Comment (projectkk2glider) [inline radio/src/mixer.cpp]: I really like the resulting simplification of expos.

File: radio/src/mixer.cpp [hunk 5]
```diff
@@ -302,7 +277,14 @@ getvalue_t getValue(mixsrc_t i)
 
   else if (i<=MIXSRC_TrimAil) return calc1000toRESX((int16_t)8 * getTrimValue(mixerCurrentFlightMode, i-MIXSRC_TrimRud));
 
-#if defined(PCBTARANIS)
+#if defined(PCBTBS)
```
Comment (projectkk2glider) [inline radio/src/mixer.cpp]: PCBTBS is unknown to me. Is is for some special board?

File: radio/src/mixer.cpp [hunk 6]
```diff
@@ -593,15 +590,22 @@ void evalFlightModeMixes(uint8_t mode, uint8_t tick10ms)
...
     uint32_t q = calc100toRESX(g_model.swashR.value);
     q *= q;
     if (v>q) {
       uint16_t d = isqrt32(v);
       int16_t tmp = calc100toRESX(g_model.swashR.value);
-      HELI_ANAS_ARRAY(ELE_STICK) = (int32_t) HELI_ANAS_ARRAY(ELE_STICK)*tmp/d;
-      HELI_ANAS_ARRAY(AIL_STICK) = (int32_t) HELI_ANAS_ARRAY(AIL_STICK)*tmp/d;
+      heliEleValue = (int32_t) heliEleValue*tmp/d;
+      heliAilValue = (int32_t) heliAilValue*tmp/d;
     }
```
Comment (projectkk2glider) [inline radio/src/mixer.cpp]: I can not comprehend why we do the same Ail Ele square summ limiting here and in evalInputs(). Wouldn't it be enough if we just did it here?

Comment (bsongis) [inline radio/src/mixer.cpp]: I don't know as well, I just kept the original algorithm (at least I hope so)

**Comment by projectkk2glider** [comment]
Ok, looks good, except the comment about double limiting. Maybe that piece of code should be checked separately in another issue to determine if it is necessary or not.

**Comment by bsongis** [comment]
Then yes, we press the button and ask for tests from heli guys?

**Comment by projectkk2glider** [comment]
Yes, do the merge and create an issue for double limiting.

**Pull request merged**